From e8ae6fb265dc6d1028dd8929657ef9afb364af3f Mon Sep 17 00:00:00 2001
From: Kevin Darbyshire-Bryant <ldir@darbyshire-bryant.me.uk>
Date: Tue, 21 Apr 2020 10:46:33 +0100
Subject: [PATCH] set a dscp maybe!

Signed-off-by: Kevin Darbyshire-Bryant <ldir@darbyshire-bryant.me.uk>
---
 src/network.c | 8 +++++++-
 1 file changed, 7 insertions(+), 1 deletion(-)

diff --git a/src/network.c b/src/network.c
index 4bada37..b58ea76 100644
--- a/src/network.c
+++ b/src/network.c
@@ -678,6 +678,7 @@ static int make_sock(union mysockaddr *addr, int type, int dienow)
 {
   int family = addr->sa.sa_family;
   int fd, rc, opt = 1;
+  int tos = IPTOS_CLASS_CS6;
   
   if ((fd = socket(family, type, 0)) == -1)
     {
@@ -714,13 +715,18 @@ static int make_sock(union mysockaddr *addr, int type, int dienow)
       
       return -1;
     }	
+
+  if (family == AF_INET6)
+	rc = setsockopt(fd, IPPROTO_IPV6, IPV6_TCLASS, &tos, sizeof(tos));
+  else
+	rc = setsockopt(fd, IPPROTO_IP, IP_TOS, &tos, sizeof(tos));
   
   if (setsockopt(fd, SOL_SOCKET, SO_REUSEADDR, &opt, sizeof(opt)) == -1 || !fix_fd(fd))
     goto err;
   
   if (family == AF_INET6 && setsockopt(fd, IPPROTO_IPV6, IPV6_V6ONLY, &opt, sizeof(opt)) == -1)
     goto err;
-  
+
   if ((rc = bind(fd, (struct sockaddr *)addr, sa_len(addr))) == -1)
     goto err;
   
-- 
2.24.2 (Apple Git-127)

